torch/nn/utils/no_init.py

```diff
@@ -0,0 +1,60 @@
+from typing import Type
+
+from ..._C import _suspend_backend_dispatch, _restore_backend_dispatch
+from ..modules import Module
+
+
+def no_init(module_cls: Type[Module], *args, **kwargs) -> Module:
+    """Constructs a module without initializing its parameters and buffers.
+
+    This function is meant to be used if the size of the specified module is big
+    and you want to inspect it without fully instantiating. For instance an
+    automated sharding system can construct a module with ``no_init()``, inspect
+    its architecture, and materialize its layers at a later time.
+
+    Internally ``no_init()`` forces all parameters and buffers of the module and
+    its sub-modules to use a meta device regardless of the requested device
+    type. The returned module can be used for inspection purposes, but has to be
+    materialized by calling its ``materialize()`` method before doing actual
+    work.
+
+    Args:
+        module_cls:
+            The type of the module to construct.
+        args:
+            The positional arguments to pass to the module's constructor.
+        kwargs:
+            The keyword arguments to pass to the module's constructor.
+
+    Returns:
+        A module instance with uninitialized parameters and buffers.
+
+    Example::
+        >>> import torch
+        >>> m = torch.nn.utils.no_init(torch.nn.Linear, 5, 1)
+        >>> m.weight
+        Parameter containing:
+        tensor(..., device='meta', requires_grad=True)
+        >>> n = m.materialize()
+        >>> n.weight
+        Parameter containing:
+        tensor([[-1.4677e+24,  4.5915e-41,  1.4013e-45,  0.0000e+00, -1.4677e+24,
+                  4.5915e-41]], requires_grad=True)
+
+    Note:
+        The ``args`` and ``kwargs`` arguments must be treated as immutable since
+        they will be used a second time during materialization.
+    """
+    _suspend_backend_dispatch()
+
+    try:
+        module = module_cls(*args, **kwargs)
+    finally:
+        _restore_backend_dispatch()
+
+    def materialize() -> Module:
+        return module_cls(*args, **kwargs)
```

Comment:
This looks great as a first step. We need some additional steps before sharing this for dog-fooding. Currently, there are at least two known use cases that has problem with large model initialization:

1. Initialize different sub-modules on different processes/devices (e.g., pipeline parallel, SageMaker model parallel)
2. Initialize different shard of the module on different processes/devices (e.g., FSDP, ZeRO, sharded Tensor)

For the first use case, we will need to 1) create the meta model, 2) trace it to get the forward graph, 3) make partitioning decisions, 4) initialize different sub-modules on different processes/devices, 5) insert communication operators accordingly. One difficulty here is that, we will also need to intercept the ctor args for submodules, and call them on a remote process. 

For the second use case, we might need some changes in `meta`/`reset_params`, as each process only owns a shard of every parameter and only needs to initialize that shard. RNG can be a problem here (not sure if that matters). Besides, the logic in `reset_params` needs to work with partial storage.

Reply:
> One difficulty here is that, we will also need to intercept the ctor args for submodules, and call them on a remote process.

@ezyang @jamesr66a is this feasible? A messy solution could be, when `device=='meta'`, the module can record the args, and expose a private API to return those args? Or maybe fx can already provide enough info to reconstruct a submodule remotely? Or, we can also pickle the meta module, communicate the pickled module to the remote process, unpickle it, and then call it's `reset_params`?